W-10590698-prEditClarifyBatchJobErrorHandling-duke (#2199)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/batch-error-handling-faq.adoc

= Handling Errors During Batch Jobs
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

== Overview

Handling errors inside a batch job is complicated by the enormous amount of data this scope is designed to handle. For example, logs from processing 1 million records become huge and unreadable, and the processing required for the logs causes major performance issues. However, logs are necessary to understand issues that occur. Without logs, debugging the failure of 30,000 records is not practical. 

== Logs of Failing Records Inside a Batch Step

> When processing a batch job instance, a processor inside a batch step may fail or raise an error. When this occurs, for example, because of corrupted or incomplete record data, Mule logs the stack traces following this logic:

Author: ```suggestion
When processing a batch job instance, a processor inside a batch step may fail or raise an error, for example, because of corrupted or incomplete record data. When such an issue occurs, Mule logs the stack traces following this logic:
```

= Handling Errors During Batch Jobs
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

== Overview

> Verbose logs for issues that occur when a batch job processes very large chunks of data can become enormous and severely impact performance. 

Author: ```suggestion
Verbose logging of issues that occur when a batch job processes very large chunks of data can become enormous and severely impact performance. 
```

> When a batch job processes very large chunks of data, verbose logs for issues that occur can become enormous and severely impact performance. 

Author: ```suggestion
Batch processing performs "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. However, when a batch job processes very large chunks of data, _verbose_ logs for issues that occur can become enormous and severely impact performance. 
```

> Verbose logging of issues that occur when a batch job processes very large chunks of data can become enormous and severely impact performance. 

Author: ```suggestion
When a batch job processes very large chunks of data, verbose logs for issues that occur can become enormous and severely impact performance. 
```

> Mule batch processing is designed to handle very large data sets and to perform "near real-time" data integration that recovers from crashes and continues processing a job from a point of failure. However, _verbose_ logs for issues that occur can become enormous and severely impact performance. 

Author: ```suggestion
Mule batch processing is designed to handle very large data sets and to perform "near real-time" data integration that recovers from crashes and continues processing a job from a point of failure. However, _verbose_ logs for issues that occur in large data sets can become enormous and severely impact performance. 
```

> Batch processing performs "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. However, when a batch job processes very large chunks of data, _verbose_ logs for issues that occur can become enormous and severely impact performance. 

Author: ```suggestion
Mule batch processing is designed to handle very large data sets and to perform "near real-time" data integration that recovers from crashes and continues processing a job from a point of failure. However, _verbose_ logs for issues that occur can become enormous and severely impact performance. 
```

> Mule batch processing is designed to handle very large data sets and to perform "near real-time" data integration that recovers from crashes and continues processing a job from a point of failure. However, _verbose_ logs for issues that occur in large data sets can become enormous and severely impact performance. 

Reviewer: What does "near real-time" mean? Where did you get this description from?

Author: @fsgonz It is a phrase that (I believe) Mariano Gonzalez used at some point and wanted to keep in the docs. Would it make sense to something like "almost real-time" or just leave that out altogether?

> Mule batch processing is designed to handle very large data sets and to perform "near real-time" data integration that recovers from crashes and continues processing a job from a point of failure. However, _verbose_ logs for issues that occur in large data sets can become enormous and severely impact performance. 

Author: ```suggestion
Mule batch processing is designed to handle very large data sets and to perform almost real-time data integration that recovers from crashes and continues processing a job from a point of failure. However, _verbose_ logs for issues that occur in large data sets can become enormous and severely impact performance. 
```

> To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. 

Author: ```suggestion
To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but is feasible for cases that involve smaller data sets. 
```

> To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but is feasible for cases that involve smaller data sets. 

Author: ```suggestion
To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful for debugging and feasible for some cases that involve smaller data sets. 
```

> To limit this impact, Mule uses INFO-level log messaging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. For example, batch processing recovers from crashes and continues processing the job from the point it "left off" when performing "near real-time" data integration. 

Author: ```suggestion
To limit this impact, Mule uses INFO-level log messaging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. For example, you might use batch processing for its "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. 
```

> To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful for debugging and feasible for some cases that involve smaller data sets. 

Author: ```suggestion
To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful for debugging and is feasible for some cases that involve smaller data sets. 
```

> To limit this impact, Mule uses INFO-level log messaging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. For example, you might use batch processing for its "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. 

Author: ```suggestion
To limit this impact, Mule uses INFO-level log messaging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. For example, you might use batch processing because it performs "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. 
```

> To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. For example, you might use batch processing because it performs "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. 

Author: ```suggestion
To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. 
```

> To limit this impact, Mule uses INFO-level log messaging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. For example, you might use batch processing because it performs "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. 

Author: ```suggestion
To limit this impact, Mule uses INFO-level logging by default, as described in <<default_logging>>. For cases in which you require more verbose log messages, you can change the mode to DEBUG. This mode is helpful not only for debugging but for some cases that involve smaller data sets. For example, you might use batch processing because it performs "near real-time" data integration, which recovers from crashes and continues processing a job from a point of failure. 
```

> The following property is available for setting the logging mode:

Reviewer: Why this general specification about logging is needed in batch?

Author: @fsgonz I believe it was added (to the original) because the INFO mode summaries do not capture the info about records that someone might want to see. See in the steps (unedited): "Even if all records raise the same error, the messages being processed would probably contain specific information related to those records. For example, if I'm pushing leads to my Salesforce account and one record fails because the lead was already uploaded, another repeated lead would have different record information, but the error is the same."

Author: Also, there are steps on this page (unedited) that describe the default process in detail, I believe because the default logging behavior might be unexpected.

[source,text,linenums]
----
log4j.logger.com.mulesoft.module.batch=DEBUG
----

> Avoid using the DEBUG mode in a production environment when processing very large datasets. 

Author: ```suggestion
Avoid using the DEBUG mode in a production environment when processing very large data sets. 
```

[[default_logging]]
== Logs of Failing Records Inside a Batch Step

> When processing a batch job instance, a processor inside a batch step can fail or raise an error, for example, because of corrupted or incomplete record data. By default, Mule logs the stack traces when such an issue occurs by following this logic:

Author: ```suggestion
When processing a batch job instance, a processor inside a batch step can fail or raise an error, for example, because of corrupted or incomplete record data. By default, Mule uses the INFO log level to log stack traces according to the following logic when issues occur:
```

> When processing a batch job instance, a processor inside a batch step can fail or raise an error, for example, because of corrupted or incomplete record data. When such an issue occurs, Mule logs the stack traces following this logic by default:

Author: ```suggestion
When processing a batch job instance, a processor inside a batch step can fail or raise an error, for example, because of corrupted or incomplete record data. By default, Mule logs the stack traces when such an issue occurs by following this logic:
```

> When processing a batch job instance, a processor inside a batch step can fail or raise an error, for example, because of corrupted or incomplete record data. By default, Mule uses the INFO log level to log stack traces according to the following logic when issues occur:

Reviewer: Again we get this information about the log level? Was this a requirement from someone?

Author: This page has been around for some time. Apparently, it is used by the training team to explain batch processing logs. It was training that raised the issue.

. Mule gets the exception's full stack trace.

> . Mule strips that stack trace from error messages. +

Author: ```suggestion
. Mule strips the stack trace from error message in the log. +
```

== Overview

> Handling errors inside a batch job is complicated by the enormous amount of data this scope is designed to handle. For example, logs from processing 1 million records become huge and unreadable, and the processing required for the logs causes major performance issues. On the other hand, logging too little would make it impossible for you to know what went wrong. And considering that if out of those 1 million records, 30 thousand of them failed, not knowing what’s wrong with them can turn your debugging task very problematic. This trade-off is not simple to overcome.

Author: ```suggestion
Handling errors inside a batch job is complicated by the enormous amount of data this scope is designed to handle. For example, logs from processing 1 million records become huge and unreadable, and the processing required for the logs causes major performance issues. However, logs are necessary to understand issues that occur. Without logs, debugging the failure of 30,000 records is not practical. 
```

************************************************************************************************************************
----
+

> Here, the first step failed ten times, and the second failed nine.

Reviewer 2: I don't think the second comma is needed, but probably fine.

Author: next time, thx
